### PR TITLE
HOTT-1599: Handle deletions on the measure mapper

### PR DESCRIPTION
**Pull request by willfish**
### Jira link

https://transformuk.atlassian.net/browse/HOTT-1586

### What?

I have added/removed/altered:

- [x] Added support for soft deletes when the measure primary is a destroy operation
- [x] Added support for missing soft deletes of the measure mapper as a primary
- [x] Added support for missing soft deletes of the measure condition mapper as a primary
- [x] Added configuration to affect mappers
- [x] Added specs to cover changes
- [x] Refactored xml entity enumeration so to avoid duplication of parse method
- [x] Added missing factory implementations
- [x] Removed some factory side effects implementations

### Why?

I am doing this because:

- This is required so we can reflect the changes that come from the CDS files properly

### CODE REVIEW

File: app/lib/cds_importer/entity_mapper/base_mapper.rb [hunk 1]
```diff
@@ -108,26 +108,22 @@ def missing_entity_filter_for(missing_entities)
...
         end
 
-        def relative_primary_key_paths_for_secondary_node
-          primary_key_paths_for_secondary_node.each_with_object({}) do |path, acc|
-            relative_path = path.sub("#{mapping_path}.", '')
-
-            acc[relative_path] = primary_key_mapping[path]
-          end
+        def entity_composite_primary_key
+          Array.wrap(entity.primary_key)
```
Comment (alexdesi) [inline app/lib/cds_importer/entity_mapper/base_mapper.rb:112]: Nice, I did know about the wrap method. It should be about
`[ entity.primary_key ]` 
but more explicit

File: app/lib/cds_importer/entity_mapper/base_mapper.rb [hunk 2]
```diff
@@ -108,26 +108,22 @@ def missing_entity_filter_for(missing_entities)
...
         end
 
-        def primary_key_paths_for_primary_node
-          primary_node_paths = primary_key_mapping.keys - primary_key_paths_for_secondary_node
+        def database_entities_for(primary_model_instance)
+          filter = filter_for(primary_model_instance)
 
-          primary_node_paths.index_with { |path| primary_key_mapping[path] }
+          entity.where(filter).pluck(*entity.primary_key).map do |composite_primary_key|
```
Comment (alexdesi) [inline app/lib/cds_importer/entity_mapper/base_mapper.rb:118]: I wonder why we need to wrap after pluck 🤔

File: app/lib/cds_importer/entity_mapper/base_mapper.rb (hunk 1 above)
Comment (willfish) [reply app/lib/cds_importer/entity_mapper/base_mapper.rb:112]: It actually behaves slightly differently:

```ruby
thing = [:foo]
Array.wrap(thing) => [:foo]

thing = :foo
Array.wrap(thing) => [:foo]

thing = [:foo]
[thing] => [[:foo]]

thing = :foo
[thing] = [:foo]
```

File: app/lib/cds_importer/entity_mapper/base_mapper.rb (hunk 2 above)
Comment (willfish) [reply app/lib/cds_importer/entity_mapper/base_mapper.rb:118]: When the primary key is singular (e.g. not a composite key), pluck returns an array with a depth of 1:

```ruby
[:foo, :bar]
```

When the primary key is a composite, pluck returns an array with a depth of 2:

```ruby
[[:foo, :baz], [:bar, :qux]]
```
I wrap so can iterate regardless of what the pluck method returns

File: spec/lib/cds_importer/entity_mapper/measure_mapper_spec.rb [hunk 3]
```diff
@@ -56,4 +56,187 @@
...
+      it_behaves_like 'an entity mapper destroy operation', Measure
+      it_behaves_like 'an entity mapper destroy operation', FootnoteAssociationMeasure
+      it_behaves_like 'an entity mapper destroy operation', MeasureComponent
+      it_behaves_like 'an entity mapper destroy operation', MeasureCondition
+      it_behaves_like 'an entity mapper destroy operation', MeasureConditionComponent
+      it_behaves_like 'an entity mapper destroy operation', MeasureExcludedGeographicalArea
+      it_behaves_like 'an entity mapper destroy operation', MeasurePartialTemporaryStop
+    end
+
+    context 'when there are missing secondary entities to be soft deleted' do
```
Comment (alexdesi) [inline spec/lib/cds_importer/entity_mapper/measure_mapper_spec.rb:216]: I don't understand this test (and the other with "... to be soft deleted" ,
What is the feature in this test that makes me think that the missing secondary entities have to be soft deleted?

File: spec/support/shared_examples/an_entity_mapper_missing_entity_destroy_operation.rb [hunk 4]
```diff
@@ -23,17 +18,17 @@
...
-
-      expect(after_import_primary_keys).to include(*before_import_primary_keys)
+    # rubocop:disable RSpec/MultipleExpectations
+    it "does not hide the missing xml nodes from the #{model} view" do
+      if has_hard_deletes
+        expect { entity_mapper.import }.to change_view_count.by(-1)
+      else
+        expect { entity_mapper.import }.not_to change_view_count
+      end
     end
```
Comment (alexdesi) [inline spec/support/shared_examples/an_entity_mapper_missing_entity_destroy_operation.rb:28]: it looks this should be split into two different contexts:
with has hard deletes
and without hard deletes

**Review by alexdesi — APPROVED** [review]
it looks good to me ✅ 
just a couple of minor comments,
the tests with conditional at the end sounds are a bit strange

File: spec/support/shared_examples/an_entity_mapper_missing_entity_destroy_operation.rb (hunk 4 above)
Comment (willfish) [reply spec/support/shared_examples/an_entity_mapper_missing_entity_destroy_operation.rb:28]: Sadly that's not actually possible

Comment (willfish) [reply spec/support/shared_examples/an_entity_mapper_missing_entity_destroy_operation.rb:28]: I'll be getting rid of this in like 2 weeks

File: spec/lib/cds_importer/entity_mapper/measure_mapper_spec.rb (hunk 3 above)
Comment (willfish) [reply spec/lib/cds_importer/entity_mapper/measure_mapper_spec.rb:216]: Added some context that should help :smile:

**Pull request merged**